### ISSUE TITLE
DNN Support matrix

### ISSUE THREAD

**Issue by tompollok** [opencv/opencv#17591]
Just an idea: 

For Mesa there is a matrix that shows the OpenGL support by different vendors/drivers
https://mesamatrix.net/

I dont know how complicated this would be to make it for opencv, but it would for sure help with making Layer Support transparent for CPU OPENCL and CUDA etc. mesamatrix is opensource.

**Comment by dkurt** [comment]
There is Vulkan support already.

**Comment by tompollok** [comment]
What do you mean? Vulkan backend for DNN?

**Comment by dkurt** [comment]
@tompollok, https://github.com/opencv/opencv/pull/12703

https://docs.opencv.org/master/d6/d0f/group__dnn.html#gga186f7d9bfacac8b0ff2e26e2eab02625a91dbe681fc3246a2bd36b04e7b65122f,

**Comment by tompollok** [comment]
Yes i know. I mean this was just an idea request to adapt mesamatrix to an opencvdnn_matrix so that everybody can see which layers or even networks are supported.

**Comment by dkurt** [comment]
Please clarify once again. What is Mesa and the Matrix? Is that a library or what? What should OpenCV community do with that? Integrate as a compute backend for dnn? This way it would be nice to see some research with comparison against existing backends.

**Comment by tompollok** [comment]
See wikipedia: https://en.wikipedia.org/wiki/Mesa_%28computer_graphics%29
Mesa, also called Mesa3D and The Mesa 3D Graphics Library, is an open source software implementation of OpenGL, Vulkan, and other graphics API specifications. Mesa translates these specifications to vendor-specific graphics hardware drivers. 

Mesamatrix just displays that API has been implemented for which graphics vendor. Its FOSS and it uses a simple text file that can be added to the opencv and then the website could be hosted through opencv.org?

I just created an example that is not complete just so that you can see what i have in mind:
![grafik](https://user-images.githubusercontent.com/4457821/86357393-0d131400-bc6e-11ea-8fbd-9cc800eca422.png)

As you can see it gives the users a "big picture" more easily of whats available and whats not. Instead of the date we can put the OpenCV Version since which it is available? I just thought that this could be useful. What do you think?

**Comment by dkurt** [comment]
I just not sure that it's really helpful. Nothing but empirical experiment can show better picture. For specific topolgy user may call [dumpToFile](https://docs.opencv.org/master/db/d30/classcv_1_1dnn_1_1Net.html#a4f7d0d1e650571e11d988d07f7ed5a7d) method to see which layers are executed on which devices.

**Comment by tompollok** [comment]
This matrix doesnt show the topology, just whether a layer has been implemented or whether a specific network runs on a backend. Of course it doesnt show if it runs completely for example on CUDA backend.
Even if empiricly tested networks are displayed here, for example with a direct download link in the network name, it just shows what currently works. Even if it is just the CPU backend. Anyways, this was just an idea proposal.

**Comment by dkurt** [comment]
The problem is that there is no list of just layers which are supported or not on specific target - there are multiple configurations in very layer and device features so the same layer may work or not on it.

**Comment by YashasSamaga** [comment]
I want to voice my support for this issue. It's quite cumbersome to track new layers and what each backend supports. I think it would help to have a support matrix doc in `modules/dnn/doc` that is updated every time a new layer or support for new configuration is added.

It's useful for end-users as well as contributors (it helps track what features are missing in the CUDA backend for example).

**Comment by alalek** [comment]
Main question is maintenance of this stuff.
We need **automatic scripts** to keep/validate that up-to-date (experience shows that "updated every time" doesn't really work).

What should be be done:
- prepare DNN test cases for each "cell" (or adapt existing tests to avoid duplicates).
- need to track "fallback" mode somehow.
- prepare script which parses tests results (`--gtest_output=xml:result.xml`) for different cases and provides such aggregated output. Some mapping `cell <=> test case` should be prepared near such script.

BTW, Markdown is not a best option for rendering information like that, but we can start from this and upgrade later.
